### ISSUE TITLE
qrexec-daemon startup failed: Connection to the VM failed

### ISSUE THREAD

**Issue by straussmaus** [QubesOS/qubes-issues#6236]
**Qubes OS version**
4.0.3

**Affected component(s) or functionality**
Qube based on a debian10 template.

**Brief summary**
When I start this VM, I get a popup with the error:

```
qrexec-daemon startup failed: Connection to the VM failed
```

There is no information with this warning.

**How Reproducible**
Happens with this Qube, and not (yet) discovered on others.  I have recently restored from backup, and not tried this Qube since restore.

**To Reproduce**

1. Restore qube based on a template (debian 10)

2. Try to start

3. See error (Above)

**Expected behavior**
The VM should start, or at least the error should tell me what logs to look at for more info.

**Actual behavior**
See above.

**Additional context**
The restored system has a fresh debian10 template.  However since this is a Qube based on a template (the old debian10) I don't think that could be the problem.  Could it?  Another Qube based on debian10 is working fine on the restored system (after also being restored).

**Solutions you've tried**

- right-click start

- start with application finder

- reboot

- start by running a command

- `virsh -c xen:/// console <qube>`  (Result: `error: The domain is not running`)

- `qrexec-client-vm <qube>` (Result: `sudo: qrexec-client-vm: command not found`  ???)

- and:

```
qrexec-run-vm
base: qrexec-run-vm: command not found
[user@dom0 ~]$ sudo !!
sudo qrexec-run-vm
sudo: qreec-run-vm: command not found
```

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**
I searched for the error message and many issues came up.  I didn't see one that seemed similar but I didn't look through them all (there were a lot).

**Comment by marmarek** [comment]
Take a look `/var/log/xen/console/guest-<qube>.log`

**Comment by straussmaus** [comment]
Thank you.  Here is some output ( tail -n 30 ) from that file:

```

[    4.582464] EXT4-fs (xvdd): mounting ext3 file system using the ext4 subsystem
[    4.592578] EXT4-fs (xvdd): mounted filesystem with ordered data mode. Opts: (null)
mount: mount point /sysroot/lib/modules does not exist
mount: /sysroot not mounted or bad option

       In some cases useful info is found in syslog - try
       dmesg | tail or so.
switch_root: failed to mount moving /sysroot to /: Invalid argument
switch_root: failed. Sorry.
[    5.407686] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
[    5.407686] 
[    5.407717] CPU: 0 PID: 1 Comm: switch_root Not tainted 4.19.155-1.pvops.qubes.x86_64 #1
[    5.407738] Call Trace:
[    5.407753]  dump_stack+0x66/0x8b
[    5.407767]  panic+0xe4/0x252
[    5.407779]  do_exit+0xa64/0xc20
[    5.407792]  ? __do_page_fault+0x278/0x4f0
[    5.407804]  do_group_exit+0x3a/0xa0
[    5.407816]  __x64_sys_exit_group+0x14/0x20
[    5.407828]  do_syscall_64+0x5b/0x190
[    5.407841]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[    5.407857] RIP: 0033:0x70a334224bd8
[    5.407868] Code: 00 00 be 3c 00 00 00 eb 19 66 0f 1f 84 00 00 00 00 00 48 89 d7 89 f0 0f 05 48 3d 00 f0 ff ff 77 21 f4 48 89 d7 44 89 c0 0f 05 <48> 3d 00 f0 ff ff 76 e0 f7 d8 64 41 89 01 eb d8 0f 1f 84 00 00 00
[    5.407916] RSP: 002b:00007ffeb6ea7988 EFLAGS: 00000246 ORIG_RAX: 00000000000000e7
[    5.407938] RAX: ffffffffffffffda RBX: 0000000000000001 RCX: 000070a334224bd8
[    5.407958] RDX: 0000000000000001 RSI: 000000000000003c RDI: 0000000000000001
[    5.407978] RBP: 000070a334514860 R08: 00000000000000e7 R09: ffffffffffffff98
[    5.407999] R10: 0000000000000003 R11: 0000000000000246 R12: 000070a334514860
[    5.408019] R13: 000070a334519c00 R14: 0000000000000000 R15: 0000000000000000
[    5.408260] Kernel Offset: disabled
```

**Comment by marmarek** [comment]
Do you see anything earlier about /sysroot failed mount? It should be xvda3 device.

**Comment by straussmaus** [comment]
Only:

```
mount: mount point /sysroot/lib/modules does not exist
mount: /sysroot not mounted or bad option
switch_root: failed to mount moving /sysroot to /: Invalid argument
```
Doing: 

`less <qube>.log | grep 'xvda3'`

turns up no matches.

**Comment by marmarek** [comment]
Anything else about xvda? Partitions list at least?

**Comment by straussmaus** [comment]
```
[    5.266135] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    3.453663] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    1.479638] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    3.982215] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    3.361823] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    2.564520] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    1.668605] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    3.650483] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    2.072430] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,
[    3.931855] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
Waiting for /dev/xvda* devices...
mount: wrong fs type, bad option, bad superblock on /dev/xvda,

```

**Comment by straussmaus** [comment]
That's everything that matches `less <qube>.log | grep 'xvda'

**Comment by marmarek** [comment]
Something is wrong with the root volume of that template, there should be:
```
[    1.866735] blkfront: xvda: flush diskcache: enabled; persistent grants: enabled; indirect descriptors: enabled;
[    1.891535]  xvda: xvda1 xvda2 xvda3
```
Is any other qube on _this very template_ working? Can you start the template itself? Is template correctly set (check `qvm-prefs <qube> template`)? Maybe try changing to another template and back.

**Comment by straussmaus** [comment]
This qube is *based on* the debian10 template, but it doesn't use it because it's a based-on, or standalone.  So, no, there are no others that use the same base.  It was working before the backup and restoration.  Other qubes based on debian10 (also based-on, standalone) are working after the restore.

**Comment by marmarek** [comment]
Ah, standalone, ok. Did you get any errors about it during restore? You can also check if its disks were restored at all:
```
qvm-ls -d <qube>
```
You should see positive numbers in "PRIV-CURR" and "ROOT-CURR". If any is 0, then it wasn't restored.

**Comment by straussmaus** [comment]
The restore said "Finished Successfully" which indicates no errors. (success != failure)

However, ROOT-CUR is 0.

The backup also gave a success message.

So perhaps the bug is in the messaging?  Saying successful when there are errors?

Could this be so, or is there another possible source for this error?

**Comment by marmarek** [comment]
Possibly, but in that case, you should get also some other message about issue during restore. Make sure also that you are looking at the right qube - if it existed before restoring, the restored one may have a different name (like `qube1`) depending on the options you've chosen - and also you'd get that info in restore summary just before confirming it.

**Comment by straussmaus** [comment]
I don't know what messaging happened during restore.  It's a 12 hour process to restore, and I come back the next day and if I see "FINISHED SUCCESSFULLY" I assume it was successful, because that's what it says.  I do not examine details of a "SUCCESSFUL" restore.

The Qube definitely did not exist before restoring because I restored onto a new installation of 4.0.3.

So here are my questions:

1 - You're confirming that the backup and restore can say "success" yet produce this obvious failure?  

2 - Can you think of something else this could be?

3 - Is there a solution besides trying another restore, and if that fails trying to back up and then restore again?

**Comment by marmarek** [comment]
I'd try to restore this particular qube again.

**Comment by straussmaus** [comment]
Okay the restore worked.  I wish it didn't because now I can't trust any restoration of Qubes until this is cleared up.

The restoration said "FINISHED SUCCESSFULLY" (or something like that) and yet, this problem required another restore.  That means that the first one didn't finish successfully.

I have about 70 Qubes that I back up and restore.  I can't check all of them shortly after every restoration.  And anyway, a basic requirement of messaging is that when there is a failure, I should not explicitly read the word "SUCCESS", right?

How can I help you get to the bottom of this?

**Comment by marmarek** [comment]
Ok, so clearly there is some issue with error reporting.
The most useful would be full output of that faulty restore, but I assume that isn't an option. Can you say something about this qube and or the backup archive? Was is large one? How long did it took to restore it on its own compared to the full restore (I try to estimate its position in the backup archive)? How big is the backup archive compared to the disk size?

**Comment by straussmaus** [comment]
> Ok, so clearly there is some issue with error reporting.

I'm glad you see that.  I am now very nervous to trust the restore and that leaves me in a bad position.

> The most useful would be full output of that faulty restore, but I assume that isn't an option.

Is it in the logs somewhere?  I can dig for the part of it that pertains to this VM.  The VM names would be too private to share here so I can't give the whole dump.

> Can you say something about this qube and or the backup archive? 

Sure, ask.

> Was is large one?

The Qube is 2048 private and 10240 system, so not really.

> How long did it took to restore it on its own compared to the full restore (I try to estimate its position in the backup archive)?

I have no idea.  It takes 12+ hours to restore so I was not watching it.  I started the restore.  Then I came back and saw "FINISHED SUCCESSFULLY" and that was all.

> How big is the backup archive compared to the disk size?

96%

**Comment by andrewdavidwong** [comment]
> This qube is _based on_ the debian10 template, but it doesn't use it because it's a based-on, or standalone.

For future reference, in Qubes terminology, "based on" means "based on a template," i.e., a VM that depends on a TemplateVM for its root filesystem, which is known as a [TemplateBasedVM](https://www.qubes-os.org/doc/glossary/#templatebasedvm). By contrast, a [StandaloneVM](https://www.qubes-os.org/doc/glossary/#standalonevm) is the opposite, namely a VM does *not* depend on any other VM for its root filesystem.

> > Was is large one?
> 
> The Qube is 2048 private and 10240 system, so not really.
> 
> > How long did it took to restore it on its own compared to the full restore (I try to estimate its position in the backup archive)?
> 
> I have no idea. It takes 12+ hours to restore so I was not watching it

12+ hours strikes me as an extremely long time for a restore. I'm guessing that the entire backup (not just that one qube) was quite large and also took an extremely long time to create in the first place. Sometimes it takes an excessively long time if the backup contains a lot of incompressible data and you don't turn off compression when creating the backup.

Large backups seem especially problematic (several reports throughout the years), but smaller backups seem to be very reliable. As a workaround for now, you might want to try breaking up your backups into groups of VMs rather than backing up the entire system at once. In addition, I suggest turning off compression for VMs that contain a lot of incompressible data. For example, a "media" VM that contains 100 GB of image and video files will not benefit from compression and could even result in a larger backup than without compression, all while taking longer to back up. (These two suggestions can work quite nicely in tandem, since you can turn on compression for a group of compressible VMs and turn it off for an incompressible group.)

**Comment by marmarek** [comment]
As for compression, personally I use `pigz` (via `qvm-backup --compress-filter`). It is massively faster, but also makes the laptop quite hot while making the backup.

**Comment by straussmaus** [comment]
@andrewdavidwong 

> For future reference, in Qubes terminology, "based on" means "based on a template,"

Thank you for the correction.

> 12+ hours strikes me as an extremely long time for a restore. I'm guessing that the entire backup (not just that one qube) was quite large and also took an extremely long time to create in the first place.

Given the size, and the media (HDD) and medium (USB 2.0) I think it's reasonable. I don't think we should look here for the source of the problem.

> Sometimes it takes an excessively long time if the backup contains a lot of incompressible data and you don't turn off compression when creating the backup.

Funny, I never noticed the compression checkbox until just now when I read this, and looked for it.  I did not know any compression was being used, and I don't see any value to compressed backups, and I will try turning it off next time - however there is still a problem to solve here first.

> In general, large backups seem especially problematic 

Then what info can I report, or what system of logging info can you create, so that we can get to the bottom of this?  My backup is under 1TB and I think that these days, anything under 1 or 2TB should be considered "not large" and should be reliable, don't you agree?

> As a workaround for now, you might want to try breaking up your backups into groups of VMs rather than backing up the entire system at once.

Thank you but this misses the point that I asked @marmarek about.  You're saying smaller backups might be more reliable.  But I have no (inline) indication that the backup or restore was bad this time, and I still do not know which of my qubes might still have problems.  And until there is error messaging that tells the user when a restore is bad, a smaller restore will not solve this problem.  I would still be wondering if I can trust my restore.  (It is not practical to check every VM thoroughly.  That would take days, maybe weeks.)

> For example, a "media" VM... will not benefit from compression

Thank you but I have no "media" qubes, and any small videos or photos I have are just incidental to other work.

But I will turn off compression before I do another backup.

However I could really use your and @marmarek's help determining what the source of this problem is because right now, I have a backup archive that might or might not be reliable, right?  And that means I can't wipe my drive and reinstall.  Should I do a fresh backup?  Is there some way to fix the error messaging, so that if I do this I can know if it is any better?

@marmarek

> As for compression, personally I use pigz (via qvm-backup --compress-filter). It is massively faster, but also makes the laptop quite hot while making the backup.

Thank you but I don't see much value to compressed backups.  Space is at a premium for a fast, daily-use SSD on the laptop.  But larger USB hard drives are cheap.  So I think I will turn compression off from now on.

The issue here at this point, is how to know if my backup is reliable.

**Comment by DemiMarie** [comment]
I wonder if some part of our backups use 32-bit length fields.

**Comment by marmarek** [comment]
> I wonder if some part of our backups use 32-bit length fields.

I don't think so ([here](https://www.qubes-os.org/doc/backup-emergency-restore-v4/) you can find more about the format). If that would be the case, restoring again wouldn't help.

What I think has happened, was running out of disk space during restore. Similar to https://github.com/QubesOS/qubes-issues/issues/5478. The issue of running out of disk space is improved in R4.1 already. But apparently we may still have some error reporting issue (although in #5478 error was reported correctly).

**Comment by andrewdavidwong** [comment]
> My backup is under 1TB and I think that these days, anything under 1 or 2TB should be considered "not large" and should be reliable, don't you agree?

Of course. I wasn't suggesting that large backups being unreliable isn't a problem. Rather, I was pointing out that it's been a known problem for a long time and providing some temporary workarounds, given the reality that the underlying cause of large backup unreliability has remain undiscovered for a long time and is therefore unlikely to be fixed immediately (though not for lack of trying).

> Thank you but this misses the point that I asked @marmarek about. You're saying smaller backups might be more reliable. But I have no (inline) indication that the backup or restore was bad this time, and I still do not know which of my qubes might still have problems. And until there is error messaging that tells the user when a restore is bad, a smaller restore will not solve this problem. I would still be wondering if I can trust my restore. (It is not practical to check every VM thoroughly. That would take days, maybe weeks.)

No, I am understanding your point and providing you with a course of action that can allow you to move forward in spite of it for now. I understand and agree that restoration false positives are very bad. By providing you with a temporary workaround, I am pointing out a way to avoid that problem. Since smaller backups are not affected by the restoration false positive problem, smaller backups allow you to continue to use backups without being affected by the problem.

By the way, I'm curious whether you've tried the [Verify backup integrity, do not restore the data](https://www.qubes-os.org/doc/backup-restore/) option to test restoring the bad backup and whether it also falsely reported success.

>> For example, a "media" VM... will not benefit from compression
> Thank you but I have no "media" qubes, and any small videos or photos I have are just incidental to other work.

That was just an example (which is why I said "For example"). What matters is that the data is incompressible.

> But I will turn off compression before I do another backup.

Yes, it may also help even with compressible data, if the problem has something to do with the large size of the backup.

> However I could really use your and @marmarek's help determining what the source of this problem is because right now, I have a backup archive that might or might not be reliable, right? And that means I can't wipe my drive and reinstall. Should I do a fresh backup? Is there some way to fix the error messaging, so that if I do this I can know if it is any better?

Like you, I lack he technical expertise to solve the underlying problem or fix the error messaging. That's not my role in this project. However, as a fellow Qubes user with some experience, I can offer what I would personally do: I would divide my VMs into smaller groups (e.g., 100-200 GB in size or smaller), then back up each group separately with compression off. Then I would use the aforementioned verification option to test restore each backup immediately after it is created. If verification is successful, I would consider those backups good and keep multiple copies in safe places.

**Comment by straussmaus** [comment]
@marmarek 

> What I think has happened, was running out of disk space during restore.

I have the same feeling, but the question remains: how can I check which of the qubes on my current backup are good and which are not, without going through them all in detail (which would take days or weeks)?

@andrewdavidwong 

> No, I am understanding your point and providing you with a course of action that can allow you to move forward in spite of it for now.

Thank you for offering a solution but it is not a solution to this problem.  This problem is that the integrity of the existing backup archive can not be verified.  

You suggest doing a new backup, in smaller blocks.  But until there is a method of verifying the integrity (such method I could just as easily apply to the existing archive), I will not know if the new archive(s) are good either.  So I will still be in the position of being unable to proceed.

This is why I keep asking @marmarek, etc. for a method of checking the backup, or perhaps a restore log to check which might indicate the source of the problem with the Qube that started this issue.  If I found a specific error (ex: "out of disk space" relating to that particular qube) then I would feel more confident that the backup was good, and that the problem is only in the restore under the condition of limited disk space.

> By the way, I'm curious whether you've tried the Verify backup integrity, do not restore the data option to test restoring the bad backup and whether it also falsely reported success.

No.  I thought about this but without doing an actual restore, it would not be possible to know if it "falsely reported success".  This process would only give new information if it reports failure, which seems unlikely since the full restore, which failed, reported success.

**Comment by DemiMarie** [comment]
> @marmarek
> 
> > What I think has happened, was running out of disk space during restore.
> 
> I have the same feeling, but the question remains: how can I check which of the qubes on my current backup are good and which are not, without going through them all in detail (which would take days or weeks)?

What I suspect is that the backups are all in fact good, but that the restore failed because there was not enough free disk space.

One way to test this theory is to install more storage into your system.  If the backup then succeeds, then lack of disk space was the problem.  I have no idea how much disk space is needed, however.

Another approach is to restore qubes a few at a time, rather than installing them all in one go.  There is probably an optimal order, but I am not sure what it is.  @marmarek do you have any ideas?

**Comment by straussmaus** [comment]
@DemiMarie

Thank you for your suggestions.  But.

> What I suspect is that the backups are all in fact good, but that the restore failed because there was not enough free disk space.

I suspect this also.  But before I risk all of my data in what could be a life changing mistake, I would like to be sure.


> One way to test this theory is to install more storage into your system.

How would this help me test?  You recall that I did not see any errors when I did the restore on the current system.  And you see that it is not practical to check the integrity of each of the qubes individually.  So even if it could be practical to acquire a larger disk, it would not be helpful.  I would do the restore as I did last time, it would say "FINISHED SUCCESSFULLY" as it did last time, and I would continue to wonder if it really was successful, or if I would soon discover some critical piece of data is missing.

> Another approach is to restore qubes a few at a time, rather than installing them all in one go. 

This is what Andrew suggested.  My reply to you is the same as to him: it does not change the situation.  Since the restore process did not give any errors, even a successful bit-bit-bit restoration would not indicate success just as the monolithic restore did not.  Everyone keeps suggesting this.  Am I missing something?

We all have the idea that the problem came from the archive being within 95% of the disk capacity.

How do we test *this* idea?  What logs can I check to get me closer?

If I knew that this qube was the last one to be restored, that would lend support to this idea.

If I knew that this qube showed errors internally despite the process saying "FINISHED SUCCESSFULLY" and I could find information on this error (meaning, it's not vague and unhelpful) then that would be good too.

But there is no way that simply doing this process again can possibly solve this problem, because there would be no way to verify that it worked that time any better than this time.  Unless there are logs to look at, in which case please tell me so I can look at them now with this restore and save the trouble of acquiring new hardware and running a new backup, etc.

**Comment by unman** [comment]
On Sat, Nov 28, 2020 at 08:58:09PM -0800, straussmaus wrote:
> @DemiMarie
> 
> Thank you for your suggestions.  But.
> 
> > What I suspect is that the backups are all in fact good, but that the restore failed because there was not enough free disk space.
> 
> I suspect this also.  But before I risk all of my data in what could be a life changing mistake, I would like to be sure.
> 

There's a level of hyperbole here that's unwarranted.
If the backup is good, then there is no risk to your data. 
If the backup is bad, then the tests you have been invited to do will
make no difference.

> 
> > One way to test this theory is to install more storage into your system.
> 
> How would this help me test?  You recall that I did not see any errors when I did the restore on the current system.  And you see that it is not practical to check the integrity of each of the qubes individually.  So even if it could be practical to acquire a larger disk, it would not be helpful.  I would do the restore as I did last time, it would say "FINISHED SUCCESSFULLY" as it did last time, and I would continue to wonder if it really was successful, or if I would soon discover some critical piece of data is missing.
> 
> > Another approach is to restore qubes a few at a time, rather than installing them all in one go. 
> 
> This is what Andrew suggested.  My reply to you is the same as to him: it does not change the situation.  Since the restore process did not give any errors, even a successful bit-bit-bit restoration would not indicate success just as the monolithic restore did not.  Everyone keeps suggesting this.  Am I missing something?

I think you *are* missing something.
You are focussed on the absence of error reporting.
Andrew and DemiMarie are focussed on the issue of identifying *if*
there was an error, and what may have caused it.
It's also important to identify if the error lies in the backup or the
restore process. The tests you have been asked to perform would help
with that.
Once this has been established then it will be worth considering where,
and how, error reporting may be useful and best implemented.

> 
> We all have the idea that the problem came from the archive being within 95% of the disk capacity.

I dont believe this has been demonstrated yet.

> 
> How do we test *this* idea?  What logs can I check to get me closer?

You have been offered a number of ways to test that idea - if you are
not prepared to do these tests, then a reasonable (and quicker)
alternative would be to take a clean install on a small disk, and
restore a scaled backup on to that.
It would help if you could state how many qubes were in the backup, of
what kind: the total size of the qubes to be restored, the disk size,
and free space before restore, the size of the standalone that failed.

> 
> If I knew that this qube was the last one to be restored, that would lend support to this idea.
> 
> If I knew that this qube showed errors internally despite the process saying "FINISHED SUCCESSFULLY" and I could find information on this error (meaning, it's not vague and unhelpful) then that would be good too.
> 
> But there is no way that simply doing this process again can possibly solve this problem, because there would be no way to verify that it worked that time any better than this time.  Unless there are logs to look at, in which case please tell me so I can look at them now with this restore and save the trouble of acquiring new hardware and running a new backup, etc.
> 
> 

I'm afraid that trouble shooting an apparent bug very often involves a
long and painful process of experimentation. You have the data that has
given rise to the problem - if you want someone else to work on it then
you either have to pass over the data (unlikely), or provide detailed
information so that the bug can be properly investigated.

In most cases it is best if the person who has the original issue takes
time to investigate it.

As far as logging is concerned, I think that the only relevant
information would be in /var/log/qubes/qubes.log - it may be worth your
while checking that, but the level of validation that you are looking
for is probably not going to be there.

**Comment by straussmaus** [comment]
>>  We all have the idea that the problem came from the archive being within 95% of the disk capacity.
> I dont believe this has been demonstrated yet.

You don't believe that our belief that the problem is (likely) from the archive being within 95% of the disk capacity has been demonstrated?  Have you read the thread?

From @DemiMarie:

> What I suspect is that the backups are all in fact good, but that the restore failed because there was not enough free disk space.

From @marmarek:

> What I think has happened, was running out of disk space during restore.

From me:

> I have the same feeling

If you *actually* mean that it has not yet been demonstrated that disk capacity was actually the cause of the problem, I agree.  No one said it was (see above).  That is precisely the question that must be determined to solve this issue.  It's exactly why I keep asking for a way to check logs, or otherwise verify the integrity of the backup on disk.  As soon as someone can provide these, we will be at Step 1.

> There's a level of hyperbole here that's unwarranted.

You are incorrect, and it is out of bounds to make assumptions about the value of my data, or the risk to it, or the predicament I am in until this issue can be solved.  But I assume your intentions were good so I'll move on.
Anyway I assume you would not say this in person so please do not say it online.

> I think you *are* missing something.

Okay I like where this is going...

> You are focussed on the absence of error reporting.

I don't see that.

> It's also important to identify if the error lies in the backup or the
restore process. The tests you have been asked to perform would help
with that.

Again, I ask: *how* will they do this?  Each time Andrew and DemiMarie have suggested something, I ran through the thought experiment and each time, the experiment gave no new data, and each time, I asked if I was wrong.  So if you think that the experiments that have been suggested will help then please find the point in the thought experiement where I was wrong.  They are not trivial tests to run.  And to repeat the obvious problem:  until there is a way to take a backup and know that the backup is good, any new backup taken will be suspect, so any experiment that depends on a new backup is not useful.  You see that right?  Am I missing something there?  It is not practical to open each VM and check all data, and it is not acceptable to lose mission critical data.  This leads to my questions (still unanswered): how can I verify the data is good, or how can I verify the restore is where things went bad?

> It would help if you could state how many qubes were in the backup, of
what kind: the total size of the qubes to be restored, the disk size,
and free space before restore, the size of the standalone that failed.

You don't say why this would help.  Please explain.

> the size of the standalone that failed.

From above: `The Qube is 2048 private and 10240 system`

> I'm afraid that trouble shooting an apparent bug very often involves a
long and painful process of experimentation. 

I appreciate the help but condecension is not ok, and this would not be tolerated in person so I ask you to refrain from it online.  But again I assume that your intentions are good.

>  I think that the only relevant
information would be in /var/log/qubes/qubes.log

There is nothing in this file that relates to the restoration.  The earliest entry in the log is from yesterday.  I guess it rotates?

> but the level of validation that you are looking
for is probably not going to be there.

So we are in the same place we were at before you commented, before DemiMarie commented, before Andrew commented: please provide some method of validating the data on disk, or validating the backup proces, or of finding a specific error relating to the problematic qube during restore.  

If you need more information please explain how it will help.

**Comment by DemiMarie** [comment]
> > It's also important to identify if the error lies in the backup or the
> > restore process. The tests you have been asked to perform would help
> > with that.
> 
> Again, I ask: _how_ will they do this? Each time Andrew and DemiMarie have suggested something, I ran through the thought experiment and each time, the experiment gave no new data, and each time, I asked if I was wrong. So if you think that the experiments that have been suggested will help then please find the point in the thought experiement where I was wrong. They are not trivial tests to run. And to repeat the obvious problem: until there is a way to take a backup and know that the backup is good, any new backup taken will be suspect, so any experiment that depends on a new backup is not useful. You see that right? Am I missing something there? It is not practical to open each VM and check all data, and it is not acceptable to lose mission critical data. This leads to my questions (still unanswered): how can I verify the data is good, or how can I verify the restore is where things went bad?

What happens if you use the “Verify backup integrity, do not restore the data” option?  If there is a problem in the backup, this should detect it.  The actual backup is cryptographically authenticated using your secret passphrase, so the likelihood that corruption will not be detected is negligible.

> > It would help if you could state how many qubes were in the backup, of
> > what kind: the total size of the qubes to be restored, the disk size,
> > and free space before restore, the size of the standalone that failed.
> 
> You don't say why this would help. Please explain.

This helps us (the Qubes developers) determine how likely the system was to run out of disk space during restore.  If the total size of the qubes to be restored exceeds the available free space before restore, then running out of space is almost guaranteed.  In this case, the restore is unlikely to succeed.  However, there should be an error message, and QubesOS should report which qubes failed to be restored.

Additionally, if this is the problem, you can solve the problem by purchasing additional storage, increasing the amount of space available.  Since QubesOS uses LVM2, this storage can be added to the existing pool without having to erase the data already on it.

On the other hand, if the total size of the qubes to be restored is much less than the available free space, then running out of free space should not happen.  In this case, the bug is probably somewhere else.

Finally, there is another possibility: the total amount of space could be enough, but the restore process could (due to a bug or poor algorithms) require too much free space for temporary storage.  I would not be surprised if it requires temporary (disk) storage equal to the size of the largest qube to be restored, but it should not require much more than that.

> > the size of the standalone that failed.
> 
> From above: `The Qube is 2048 private and 10240 system`

This is the amount of space that the qube is allowed to use.  QubesOS uses thin provisioning, meaning that disk space is not allocated until needed.  What we need to know is how much space the qube actually was using.

> > I think that the only relevant
> > information would be in /var/log/qubes/qubes.log
> 
> There is nothing in this file that relates to the restoration. The earliest entry in the log is from yesterday. I guess it rotates?

It does, but there might be additional info in the system journal.

Would you be willing to retry restoring the failed StandaloneVM?  This will not erase any data except for the broken VM (which presumably is not usable).

**Comment by DemiMarie** [comment]
@marmarek: what if we did the following on a test machine that holds no data?

- Create two qubes, each of which use almost half of the available storage space.
- Back up both qubes.
- Erase both qubes.
- Try to restore the backup.

If my intuition is correct, this should result in a broken restore.

**Comment by DemiMarie** [comment]
@straussmaus: if you could try to restore some of your largest qubes, without overwriting the existing ones, that would be greatly appreciated.  Please only do this if you can do so without losing any data.

**Comment by iamahuman** [comment]
I strongly believe "verify backup" option won't do anything--at least the last time when I checked the source code, it didn't do anything the full restore doesn't do (e.g. comparing uncompressed backup with current data bit-for-bit). Please correct me if I'm wrong though, I didn't thoroughly looked into it anyway.

Assuming the silent backup failure was not caused by a hardware error, I think the best _workaround_ (...ouch) for @straussmaus 's specific problem for now would be comparing hashes of volumes of restored VMs with original ones (_possibly_ from a different computer).

**Comment by unman** [comment]
On Sun, Nov 29, 2020 at 10:29:23PM -0800, straussmaus wrote:
> >>  We all have the idea that the problem came from the archive being within 95% of the disk capacity.
> > I dont believe this has been demonstrated yet.
> 
> You don't believe that our belief that the problem is (likely) from the archive being within 95% of the disk capacity has been demonstrated?  Have you read the thread?
> 
> >From @DemiMarie:
> 
> > What I suspect is that the backups are all in fact good, but that the restore failed because there was not enough free disk space.
> 
> >From @marmarek:
> 
> > What I think has happened, was running out of disk space during restore.
> 
> >From me:
> 
> > I have the same feeling
> 
> If you *actually* mean that it has not yet been demonstrated that disk capacity was actually the cause of the problem, I agree.  No one said it was (see above).  That is precisely the question that must be determined to solve this issue.  It's exactly why I keep asking for a way to check logs, or otherwise verify the integrity of the backup on disk.  As soon as someone can provide these, we will be at Step 1.
> 
> > There's a level of hyperbole here that's unwarranted.
> 
> You are incorrect, and it is out of bounds to make assumptions about the value of my data, or the risk to it, or the predicament I am in until this issue can be solved.  But I assume your intentions were good so I'll move on.
> Anyway I assume you would not say this in person so please do not say it online.
> 
> > I think you *are* missing something.
> 
> Okay I like where this is going...
> 
> > You are focussed on the absence of error reporting.
> 
> I don't see that.
> 
> > It's also important to identify if the error lies in the backup or the
> restore process. The tests you have been asked to perform would help
> with that.
> 
> Again, I ask: *how* will they do this?  Each time Andrew and DemiMarie have suggested something, I ran through the thought experiment and each time, the experiment gave no new data, and each time, I asked if I was wrong.  So if you think that the experiments that have been suggested will help then please find the point in the thought experiement where I was wrong.  They are not trivial tests to run.  And to repeat the obvious problem:  until there is a way to take a backup and know that the backup is good, any new backup taken will be suspect, so any experiment that depends on a new backup is not useful.  You see that right?  Am I missing something there?  It is not practical to open each VM and check all data, and it is not acceptable to lose mission critical data.  This leads to my questions (still unanswered): how can I verify the data is good, or how can I verify the restore is where things went bad?
> 
> > It would help if you could state how many qubes were in the backup, of
> what kind: the total size of the qubes to be restored, the disk size,
> and free space before restore, the size of the standalone that failed.
> 
> You don't say why this would help.  Please explain.
> 
> > the size of the standalone that failed.
> 
> >From above: `The Qube is 2048 private and 10240 system`
> 
> > I'm afraid that trouble shooting an apparent bug very often involves a
> long and painful process of experimentation. 
> 
> I appreciate the help but condecension is not ok, and this would not be tolerated in person so I ask you to refrain from it online.  But again I assume that your intentions are good.
> 
> >  I think that the only relevant
> information would be in /var/log/qubes/qubes.log
> 
> There is nothing in this file that relates to the restoration.  The earliest entry in the log is from yesterday.  I guess it rotates?
> 
> > but the level of validation that you are looking
> for is probably not going to be there.
> 
> So we are in the same place we were at before you commented, before DemiMarie commented, before Andrew commented: please provide some method of validating the data on disk, or validating the backup proces, or of finding a specific error relating to the problematic qube during restore.  
> 
> If you need more information please explain how it will help.
> 

I would say exactly the same to you in person.

Thought experiments aren't much use here.
We need real data to work with, and again you haven't provided it.
The log does rotate - if you look in the /var/log/qubes directory you
should find the rotated logs under qubes.log-DATE

A useful (real) experiment would be to use a much smaller disk, a clean install
of Qubes, and a backup of qubes that will fill a large percentage of the
disk.

Did you update the system before the restore?
You say you restore about 70 qubes, and it's implicit that you backed up
to an external USB device - was the reference to HDD the disk FROM
or TO which you were backing up.
This Qube is 2048 private and 10240 system
The backup is "under 1TB" - is this the size of the backup file, or the
cumulative size of the qubes you are backing up?
What is the allocated size of the qubes you are restoring, and what was
the disk capacity and free disk capacity before the restore?
What is the system configuration? 

All of this will help to identify exactly where the issue lies - is it
the size of the backup? The size in relation to the free disk space? The
number of qubes in the backup? Lack of system resources?

**Comment by unman** [comment]
On Sun, Nov 29, 2020 at 11:21:06PM -0800, Demi Marie Obenour wrote:
> @marmarek: what if we did the following on a test machine that holds no data?
> 
> - Create two qubes, each of which use almost half of the available storage space.
> - Back up both qubes.
> - Erase both qubes.
> - Try to restore the backup.
> 
> If my intuition is correct, this should result in a broken restore.
> 

Yes

**Comment by iamahuman** [comment]
Also https://www.qubes-os.org/doc/backup-emergency-restore-v4/ might help.

**Comment by andrewdavidwong** [comment]
> what if we did the following on a test machine that holds no data?
> 
> * Create two qubes, each of which use almost half of the available storage space.
> * Back up both qubes.
> * Erase both qubes.
> * Try to restore the backup.
> 
> If my intuition is correct, this should result in a broken restore.

Out of curiosity, why is that? I have no reason to doubt you, and the reason is probably obvious to many others, just not to me. I'm asking only out of curiosity and for my own edification.

**Comment by unman** [comment]
On Tue, Dec 01, 2020 at 02:44:37PM -0800, Andrew David Wong wrote:
> > what if we did the following on a test machine that holds no data?
> > 
> > * Create two qubes, each of which use almost half of the available storage space.
> > * Back up both qubes.
> > * Erase both qubes.
> > * Try to restore the backup.
> > 
> > If my intuition is correct, this should result in a broken restore.
> 
> Out of curiosity, why is that? I have no reason to doubt you, and the reason is probably obvious to many others, just not to me. I'm asking only out of curiosity and for my own edification.
> 

It's testing the hypothesis that the issue lies with restoring data
which occupies a large portion of the disk. It was said that the data
was less than 1TB, and the archive within 95% of the disk capacity.
The proposed experiment will create an example that fits the latter
condition.

**Comment by unman** [comment]
Is there any significance in the "95% of disk capacity", and what does
this mean?

What size is the disk??
What size is the root partition?
What size is the archive?
What does the disk space widget report after full restore?

**Comment by straussmaus** [comment]
> I would say exactly the same to you in person.

That would be an error.  But let's focus on the issue.

>  you look in the /var/log/qubes directory you
should find the rotated logs under qubes.log-DATE

Ah, thank you.  I do find one earlier log, but it was so much later than the restore that I did not recognize it as valuable.  And it wasn't (no errors regarding the restore of this qube) but for a different reason.  Thank you and I will know to look for this in future.

> A useful (real) experiment would be to use a much smaller disk, a clean install
of Qubes, and a backup of qubes that will fill a large percentage of the
disk.

How would this experiment run?  I say again that there were *no errors* for the backup or the restore leading to this issue.  If this holds, then I could perform the experiment you suggest and it would appear that everything went well and we would have no way of knowing if that is true.

This is why I keep asking for a way to verify the data on disk, or a log that might show an error either in backup or restore.

Why do I need to repeat this so often?  Either I am missing something, and you will please finally tell me what this secret is that you are keeping from me, or you will see that I am right and you will stop suggesting experiments that by their design can not produce new information.

It is not trivial to free up or purchase a new disk, install Qubes, calculate 95% restore capacity, and wait overnight for the install, particularly if the end result will be... what?  Another "FINISHED SUCCESSFULLY" message that may or may not be true?

> Did you update the system before the restore?

No.  It was a raw 4.0.3 install.

> You say you restore about 70 qubes, 

More or less.  

> and it's implicit that you backed up to an external USB device

Yes.  USB 2.0, at least for the restore.  I'm not sure if the backup was 2.0 or 3.0

> was the reference to HDD the disk FROM or TO which you were backing up.

TO.  The live disk was SSD and the USB backup drive was HDD.  The destination (restore) disk was also SSD, although smaller than the original live disk.

> The backup is "under 1TB" - is this the size of the backup file, or the
cumulative size of the qubes you are backing up?

I think the answer is that these are one and the same, right?  The backup utility seems to concatenate all of the qubes into one archive.

> What is the allocated size of the qubes you are restoring, and what was
the disk capacity and free disk capacity before the restore?

I'm sorry, I don't understand.  Or maybe I do?  The restore target disk was a 500GB SSD, with only a fresh installation of Qubes.  I don't recall the numbers but I compared a "df -h" to the reported size of the restored qubes and I chose not to restore certain Qubes, in order to be within capacity.  However I did not know that the backup was compressed.  Would the restore utility have told me the compressed or uncompressed size?

Once I finished the install, I had a warning that I had 4.x% disk capacity remaining (which is where I get the "95% capacity" figure, above)

5% of 500GB is 25GB right?  And the Qube that I *know* had an error is about 1/2 that capacity.  I do not know (and might go for months before learning) that others also were not restored correctly.

> What is the system configuration?

That is a broad question.  Can you be more specific?  The target was a fresh install of Qubes 4.0.3

Thank you for working with me on this.

**Comment by straussmaus** [comment]
@DemiMarie

> if you could try to restore some of your largest qubes, without overwriting the existing ones, that would be greatly appreciated.

I can.  But can you please answer the question I keep asking everyone who is suggesting something like this:  *what would this give us*?  Because if this restore goes like the last one, it will say "FINISHED SUCCESSFULLY" even if it has not finished successfully.  And it is not feasible to examine all of the data in even a small number of the largest qubes to look for things that arent there.  So what would we learn from doing this?

In fact I could even completely wipe the new/target/smaller disk and re-do the entire restore.  I still have my original data on my original disk (and will continue to have it, until this issue is resolved), so there is no risk of data loss.

Do you think it would be useful to do this?  A full repeat of the original restore?

What would we learn if we did this?

I think your guess is probably correct, that there is some unlogged weirdness about restoring a Qube when there is a small amount


```
@marmarek: what if we did the following on a test machine that holds no data?

    Create two qubes, each of which use almost half of the available storage space.
    Back up both qubes.
    Erase both qubes.
    Try to restore the backup.

If my intuition is correct, this should result in a broken restore.
@DemiMarie
DemiMarie commented 2 days ago
```

This is a good experiment, but they should be standalone qubes, and they should have data that you can verify - SHIT!

I think I just thought of the best test to run, at least so far.  It is time-consuming, but it might be necessary.  I can take a hash of each of the restored qubes and compare it to a hash on the original system, yes?  Maybe if I export it to a spreadsheet I can compare both without going bug-eyed.

Does this seems feasible?  Where do I find the qubes on dom0 please?

**Comment by straussmaus** [comment]
@imahuman

> I strongly believe "verify backup" option won't do anything--at least the last time I checked the source code, it didn't do anything the full restore doesn't do (e.g. comparing uncompressed backup with current data bit-for-bit).

Okay, at least we can stop looking in this direction.  Thank you.

> Assuming the silent backup failure was not caused by a hardware error,

Maybe it was?

> Assuming the silent backup failure was not caused by a hardware error, I think the best workaround (...ouch) for @straussmaus 's specific problem for now would be comparing hashes of volumes of restored VMs with original ones (possibly from a different computer). That way you can rest assured that restore worked correctly.

Damn.  I thought I came up with this idea myself.  But you beat me to it.

I agree this is the best solution.  It's not a great one but it might be the best one.

**Feature Request:  Qubes Backup should store the hash of the qube on source disk and compare it to the restored hash.  If they do not match, Qubes should not report "FINISHED SUCCESSFULLY" on restore.**

**Comment by DemiMarie** [comment]
> @DemiMarie
> 
> > if you could try to restore some of your largest qubes, without overwriting the existing ones, that would be greatly appreciated.
> 
> I can. But can you please answer the question I keep asking everyone who is suggesting something like this: _what would this give us_? Because if this restore goes like the last one, it will say "FINISHED SUCCESSFULLY" even if it has not finished successfully. And it is not feasible to examine all of the data in even a small number of the largest qubes to look for things that arent there. So what would we learn from doing this?

I am not super familiar with how the QubesOS backup restore process works, but I wonder if restoring a qube requires scratch space comparable to its size.  This would be the case if the backup was restored into a temporary file and then copied into the final location.  This ensures that we don’t restore a truncated volume ― unless we run out of space in the process.

We can fix this by restoring into a temporary logical volume and checking its hash before renaming it into to the final value.  That doesn’t require O(n) scratch space, and I believe it is atomic to boot.  @marmarek what do you think?

**Comment by marmarek** [comment]
It needs temporary space to extract and verify individual data chunks and only then write them to the target volume (which already is a temporary volume that is later renamed on success). The problem is that the process is highly asynchronous and it may happen that retrieving data from the archive is faster than verification + decompressing + writing to the final space - which may result in out of disk space situation.
This specific is already tracked in many related tickets already: https://github.com/QubesOS/qubes-issues/issues/3230 https://github.com/QubesOS/qubes-issues/issues/5478 and partially https://github.com/QubesOS/qubes-issues/issues/1652

**Comment by brendanhoar** [comment]
One other item to note (may not be relevant here):

If using the default storage system of lvm thin provisioning, a backup process over a set of VMs that was itself came into being via VM cloning operations would likely take up substantially more real storage space (as an uncompressed backup) and significantly more space after restore.

**Comment by straussmaus** [comment]
Can not attach USB port to Qube


These things make sense.  I still will check the hashes and post here.  But it's not something I can do easily or quickly.  I know it is critical to understanding this issue and I plan to do it.

Anyway it would help if there was a check done to ensure enough free disk space and *most important* is that the "FINISHED SUCCESSFULLY" note does not display if there are errors.

**Comment by iamahuman** [comment]
@straussmaus The quickest way to check local hashes is `sha<N>sum /dev/<pool>/vm-*-{private,root}` (where `<N>` is the version of SHA to use and `<pool>` is your LVM pool's name).

For VMs stored in backup, you can restore them (which creates VMs with prefixes or suffixes so as to not collide with existing VMs) and do the same.

**Comment by straussmaus** [comment]
@iamahuman How can I find the LVM pool's name?  I did not pay attention to the automatic partitioner during install.

**Comment by straussmaus** [comment]
Also, I don't understand `vm-*-{private,root}` please?

Okay under qubes_dom0 I see a lot of filenames like my qube names.  However for every one, there is a `vm-<something>` and also a `vm-<something>-back`.  What's that about?  And they are all symlinks to things like `../dm-150`.  Am I in the right place?

**Comment by unman** [comment]
On Sun, Dec 06, 2020 at 02:29:02AM -0800, straussmaus wrote:
> Also, I don't understand `vm-*-{private,root}` please?
> 
> Okay under qubes_dom0 I see a lot of filenames like my qube names.  However for every one, there is a `vm-<something>` and also a `vm-<something>-back`.  What's that about?  And they are all symlinks to things like `../dm-150`.  Am I in the right place?
> 

Yes, you are - these are the root and private disks, along with
backups. These are human readable, and the actual thin provisioned
volumes are under /dev
You can mount/copy/process either.

**Comment by iamahuman** [comment]
> On Sun, Dec 06, 2020 at 02:29:02AM -0800, straussmaus wrote: Also, I don't understand `vm-*-{private,root}` please? Okay under qubes_dom0 I see a lot of filenames like my qube names. However for every one, there is a `vm-<something>` and also a `vm-<something>-back`. What's that about? And they are all symlinks to things like `../dm-150`. Am I in the right place?
> Yes, you are - these are the root and private disks, along with backups. These are human readable, and the actual thin provisioned volumes are under /dev You can mount/copy/process either.

Just to clarify: the `*-back` volumes are _revisions_ of VM volumes that you can revert to using `qvm-volume revert <VM>:<volume> <revision>`. They are visible under `Available revisions (for revert):` section at the end of `qvm-volume info <VM>:<volume>`'s output. It has nothing to do with the Qubes backup utility (`qvm-backup` / Backup Qubes) this issue is about.

**Comment by straussmaus** [comment]
@iamahuman 

Okay.  I am still trying to write a little script that takes the hashes and writes them all to a file, one line at a time.  My brain is broken and the one I ran overnight failed.  I'll try again when I can.

**Comment by straussmaus** [comment]
Okay I needed two >> instead of > of course.  I was tired.  The scripts ran and I compared on a new backup on a drive of the same capacity (this is an expensive omission in the error messaging!).  Most hashes match.  However I have some questions:

- Is it true that simply opening a qube will change its hash?  That makes sense.

- What is "private" and "root" and "snap" in the names of the qubes?

- Some of them only have `private-snap` or `private-x-back`

- Sometimes the backup vs restore doesn't match, as in there are extras like `x-private-y` and `x-private-z` where y and z are random-seeming numbers

Is there a doc on what to expect here?

**Comment by straussmaus** [comment]
@iamahuman 

I think I am close to being able to close this issue, if I can confirm the above.  Do you have any information on those items?

**Comment by straussmaus** [comment]
@unman do you have any insight?

**Comment by iamahuman** [comment]
What was this issue about again? Perhaps it's a good idea to split this mega thread into multiple sub-issues before we can mark this as resolved.

Although discussions shall go to [either the mailing list or the Discourse group](https://www.qubes-os.org/support/), I think it wouldn't hurt adding a few more words here.

> * Is it true that simply opening a qube will change its hash?  That makes sense.

Yes, if the VM writes (or issue discard commands) to the disk.

> 
> * What is "private" and "root" and "snap" in the names of the qubes?

See https://www.qubes-os.org/doc/template-implementation/. It's a little bit out of date in that LVM is now preferred to file-backed loop devices, but it still provides a good overview.

> 
> * Some of them only have `private-snap` or `private-x-back`

Ephemeral volumes (e.g. root volumes of an [TemplateBasedVM](https://www.qubes-os.org/doc/glossary/#templatebasedvm)) only have `-snap`s and no revisions.

If there are `-back[-TS]` revision LVs but no corresponding original (unsuffixed) LV, I cannot think of any other scenarios than something having gone wrong. You can still use `qvm-revert` to restore the volume to any older `-back[-TS]` revisions in this case.

> 
> * Sometimes the backup vs restore doesn't match, as in there are extras like `x-private-y` and `x-private-z` where y and z are random-seeming numbers

When a VM is shut down, its current volume is copied to its `-back` volume. If an older `-back` volume already exists, Qubes renames the older volume by appending the current timestamp to its name. Thus, the numerical suffix after `-back-` in a revision's name is the timestamp at which the VM was shut down for the second time _after_ the revision was made.

> 
> 
> Is there a doc on what to expect here?

Maybe we should update https://github.com/QubesOS/qubes-doc/blob/master/user/common-tasks/volume-backup-revert.md.

**Comment by straussmaus** [comment]
This isn't a megathread.  It's an issue, alerted by the error message in the (original) title, and we are working the main hypothesis.

Summary: 

A restored qube would not start.

Restoring it again worked.

There was a success message on first restore.  *This is the first bug to fix.*

The restore archive was 95% capacity of the restore drive.

@DemiMarie and @marmarek think the disk space may have caused a restore to fail.  I agree.  Some disagreed.

They did not respond to my questions regarding how to test this.  This left the issue in the very bad position of not being able to trust any backups for any purposes.

@unman continued to suggest tests that could not even theoretically lead to more information, and did not respond to request for clarification on this subject.

The best way to test, from @iamahuman, was determined to be comparing the hashes of the qubes on each system.

I did this.

Most were the same, some were not (confirmed OK as they were opened), and some naming conventions were not clear.

A manual check was just done on some remaining qubes and I say that the archive (the same archive as the failed restore) is good.

Therefore the problem was not in the backup, but in the restore.

Therefore the issue is:

1 - Restoring from archive should not indicate success when some qubes do not restore.

2 - Qubes should warn user if there is not enough disk space to restore an archive properly.


Other questions remain, regarding "back" "snap" and your suggestion that every time a qube is shut down a duplicate is made (`When a VM is shut down, its current volume is copied to its -back volume`) however they are no longer relevant to solving this issue so not properly continued here.

I will re-title this issue with (1), create a new issue with (2), and put the error message in the body.

I will not close the issue because it is still an open bug (see 1 and 2), not user error.

**Comment by iamahuman** [comment]
Thanks! It would be appreciated to update the OP issue summary as well, since this turned out not to be a problem with qrexec or VM startup.

**Comment by andrewdavidwong** [comment]
@iamahuman raises a good point. The discussion on this issue is way too verbose, which makes it harder for the developers to keep track of what action actually needs to be taken. Anything that is not essential to the developers doing their work should be left out.

@straussmaus, it's not up to you to decide how the issue is to be diagnosed. It's up to the Qubes developers. This issue tracker is a tool to support their work, not a place for user support or answering questions. This issue (and many of the others you've opened) has demanded a disproportionate amount of time from the developers due to excessive discussion in the comments. You're also bumping this issue and tagging individual developers excessively. For these reasons, I'm going to lock this issue issue for now. (Devs, please feel free to unlock this at your discretion.)
